proposals/left-right-join-in-query-expressions.md

```diff
@@ -0,0 +1,235 @@
+> ```
+>
+> *end example*
+
+A `left join` clause followed by a query body clause:
+
+```csharp
+from «x1» in «e1»  
+left join «x2» in «e2» on «k1» equals «k2»  
+...
+```
+
+is translated into
+
+```csharp
+from * in ( «e1» ) . LeftJoin(  
+«e2» , «x1» => «k1» , «x2» => «k2» ,
+( «x1» , «x2» ) => new { «x1» , «x2» })  
+...
+```
+
+### Right join translation
+
+A `right join` clause immediately followed by a `select` clause
+
+```csharp
+from «x1» in «e1»  
+right join «x2» in «e2» on «k1» equals «k2»  
+select «v»
+```
+
+is translated into
+
+```csharp
+( «e1» ) . RightJoin( «e2» , «x1» => «k1» , «x2» => «k2» , ( «x1» , «x2» ) => «v» )
+```
+
+> *Example*: The example
+>
+> ```csharp
+> from o in orders
+> right join c in customersh on o.CustomerID equals c.CustomerID
+> select new { o?.OrderDate, o?.Total, c.Name }
+> ```
+>
+> is translated into
+>
+> ```csharp
+> (orders).RightJoin(
+>    customers,
+>    o => o.CustomerID, c => c.CustomerID,
+>    (c, o) => new { o?.OrderDate, o?.Total, c.Name })
+> ```
+>
+> *end example*
+
+A `right join` clause followed by a query body clause:
+
+```csharp
+from «x1» in «e1»  
+right join «x2» in «e2» on «k1» equals «k2»  
+...
+```
+
+is translated into
+
+```csharp
+from * in ( «e1» ) . RightJoin(  
+«e2» , «x1» => «k1» , «x2» => «k2» ,
+( «x1» , «x2» ) => new { «x1» , «x2» })  
+...
+```
+
+### Nullability of range variables
+
+A regular (inner) join introduces a range variable whose nullability follows from its source sequence. In other words, given the following code:
+
+```csharp
+from c in customersh
+join o in orders on c.CustomerID equals o.CustomerID
+select new { c.Name, o.OrderDate, o.Total }
+```
+
+... the nullability of the range variable `o` flows from `orders` (`o` is an `Order?` if `orders` is `List<Order?>`).
+
+In contrast, since `left join` returns outer elements which don't have correlated inner elements, the inner range variable it introduces is always nullable:
+
+```csharp
+from c in customersh
+left join o in orders on c.CustomerID equals o.CustomerID
+select new { c.Name, o?.OrderDate, o?.Total }
+```
+
+... in this example, `o` is an `Order?` even if `orders` is a `List<Order>`. Note that "nullable" in this context means `T?` in an unconstrained generic context; in other words, if `orders` is a sequence of value types (e.g. `List<int>`), then `o` has the same type as the elements of that sequence (`int`, not `int?`).
+
+`right join` operates in a similar way, with one important difference: the already-existing, outer range variable is made nullable, rather than the inner range variable:
+
+```csharp
+from o in orders
+right join c in customersh on o.CustomerID equals c.CustomerID
+select new { o?.OrderDate, o?.Total, c.Name }
+```
+
+In other words, if `o` was non-nullable before the `right join` operation (since `orders` is a sequence of non-nullable elements), the `right join` operation makes it nullable.
+
+## Drawbacks
```

Comment:
I would mention that this *will* work in ExpressionTrees, and will NOT call into new nodes.  It will use existing nodes, just referencing the linq LeftJoin/RightJoin methods respectively.